File: 1870-minimum-speed-to-arrive-on-time/1870-minimum-speed-to-arrive-on-time.java
class Solution {
    public int minSpeedOnTime(int[] dist, double hour) {
        int max = -2;
        for(int num : dist) max = Math.max(max, num*1000);
        int min = 1;
        double sum = 0.0;
        int k = 0;
        int lastCorr = -1;
        while(min <= max){
            k = (max - min) / 2 + min;  
            sum = 0.0;
            for(int num = 0 ; num < dist.length; ++num){
                if(num == dist.length - 1){
                    sum += (double) ((double) dist[num])/(k+0.0);
                    if(sum <= hour) lastCorr = k;
                }
                else{
                    sum += (int) dist[num]/k;
                    if(dist[num] % k != 0) sum += 1;
                    
                }
                if(sum > hour){
                    break;
                }
            }
            if(hour == sum) return k;
            else if(sum > hour){
                min = k+1;
            }
           else {
                max = k-1;
            }    
        }
        return (hour == sum) ? k : lastCorr;
    }
}

File: 1870-minimum-speed-to-arrive-on-time/1870-minimum-speed-to-arrive-on-time.cpp
class Solution {
public:
    bool canbeTravelled(vector<int>& dist, int speed, double hour){
        double time = 0;int i = 0;
        for(; i < dist.size() - 1; i++){
            if(dist[i] % speed == 0) time += (dist[i] / speed);
            else time += (dist[i]/speed) + 1;
        }
        time += (dist[i]/(double)speed);
        if(time <= hour) return true;
        return false;
    }
    
    int minSpeedOnTime(vector<int>& dist, double hour) {
        int ans = -1;
        int l = 1, r = 1000000000;
        while(l <= r){
            int mid = (l + r)/2;
            if(canbeTravelled(dist, mid, hour)){
                ans = mid; r = mid - 1;
            }else{
                l = mid + 1;
            }
        }
        return ans;
    }
};